File: arrays_and_hashing/0238_product_of_array_except_self.py
# 0238_product_of_array_except_self

class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        result = []

        prefixResult = 1
        result.append(prefixResult)

        for i in range(len(nums) - 1):
            prefixResult *= nums[i]
            result.append(prefixResult)

        postfixResult = 1

        for j in range(len(nums) - 1, 0, -1):
            postfixResult *= nums[j]
            result[j - 1] *= postfixResult

        return result

File: arrays_and_hashing/0238_product_of_array_except_self.cpp
// https://leetcode.com/problems/product-of-array-except-self/

#include <vector>

class Solution {
public:
    std::vector<int> productExceptSelf(std::vector<int>& nums) {
        std::vector<int> result(nums.size(), 1);

        int prefixResult = 1;

        for (int i = 0; i < nums.size() - 1; ++i) {
            prefixResult *= nums[i];
            result[i + 1] = prefixResult;
        }

        int postfixResult = 1;

        for (int j = nums.size() - 1; j > 0; --j) {
            postfixResult *= nums[j];
            result[j - 1] *= postfixResult;
        }

        return result;
    }
};
